src/main/java/co/cask/hydrator/salesforce/plugin/source/batch/SalesforceWideRecordReader.java

```diff
@@ -0,0 +1,238 @@
+    }
+    return result;
+  }
+
+  /**
+   * Fetches single entry map (Id -> SObjectId_value) values received from Bulk API.
+   *
+   * @param inputSplit         specifies batch details
+   * @param taskAttemptContext task context
+   * @return list of single entry Map
+   * @throws IOException          can be due error during reading query
+   * @throws InterruptedException interrupted sleep while waiting for batch results
+   */
+  private List<Map<String, String>> fetchBulkQueryIds(InputSplit inputSplit, TaskAttemptContext taskAttemptContext)
+    throws IOException, InterruptedException {
+    super.initialize(inputSplit, taskAttemptContext);
+    List<Map<String, String>> fetchedIdList = new ArrayList<>();
+    while (super.nextKeyValue()) {
+      fetchedIdList.add(super.getCurrentValue());
+    }
+    return fetchedIdList;
+  }
+
+  /**
+   * Transforms list of single entry map to array of SObject ids.
+   * <p/>
+   * Example:
+   * <ul>
+   *  <li>Expected list of map format: `List(Map(Id -> SObject_id1), Map(Id -> SObject_id2), ...)`</li>
+   *  <li>Result array: `[SObject_id1, SObject_id2, ...]`</li>
+   * </ul>
+   * @param subIds list of single entry Map
+   * @return array of SObject ids
+   */
+  private String[] getSObjectIds(List<Map<String, String>> subIds) {
+    return subIds.stream()
+      .map(Map::values)
+      .flatMap(Collection::stream)
+      .toArray(String[]::new);
+  }
+
+  /**
+   * Fetches wide object records through SOAP API.
+   *
+   * @param partnerConnection SOAP connection
+   * @param fields            SObject fields to be fetched
+   * @param sObjectName       SObject name
+   * @param sObjectIds        SObject ids to be fetched
+   * @return fetched SObject array
+   */
+  private SObject[] fetchPartition(PartnerConnection partnerConnection, String fields, String sObjectName,
+                                   String[] sObjectIds) {
+    try {
+      return partnerConnection.retrieve(fields, sObjectName, sObjectIds);
+    } catch (ConnectionException e) {
+      LOG.trace("Fetched SObject name: '{}', fields: '{}', Ids: '{}'", sObjectName, fields,
+                String.join(",", sObjectIds));
+      throw new RuntimeException(String.format("Cannot retrieve data for SObject '%s'", sObjectName), e);
+    }
+  }
+
+  /**
+   * Extracts value from XmlObject field. Reference type fields extracted recursively.
+   * <p/>
+   * Example:
+   * <ul>
+   *  <li>SObject Opportunity simple field `Name`: `XmlObject{name=Opportunity, value=null, children=[
+   *  XmlObject{name=Name, value=value1, children=[]}]}`
+   *  <br/>
+   *  <b>Result:</b> `value1`</li>
+   *  <li>SObject Opportunity reference field `Campaign.Name`: `XmlObject{name=Opportunity, value=null, children=[
+   *    XmlObject{name=Campaign, value=null, children=[XmlObject{name=Name, value=value2, children=[]}]}]}`
+   *  <br/>
+   *  <b>Result:</b> `value2`</li>
+   * </ul>
```

Comment:
what about this example: XmlObject{name=Opportunity, value=null, children=[XmlObject{name=Campaign, value=null, children=[XmlObject{name=Name, value=value2, children=[]}], XmlObject{value=null, children=[XmlObject{name=Name, value=value3, children=[]}]]}? Should result be value2 or value3 or is this not a possible example?

Reply:
Salesforce doesn't allow duplicate fields in query so this case is not possible.

Reply:
Ok - but what about XmlObject{name=Opportunity, value=null, children=[XmlObject{name=Campaign, value=null, children=[XmlObject{name=Name, value=value2, children=[]}], XmlObject{name=Name, value=value3, children=[]}]}? Or is this not possible too? Perhaps commenting/adding preconditions here would be helpful